src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -111,6 +111,209 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_OFI_win_init_sep
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+/* Activate per-window EP/counter using scalable ep. */
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_win_init_sep(MPIR_Win * win)
+{
+    int i, mpi_errno = MPI_SUCCESS;
+    struct fi_info *finfo;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_WIN_INIT_SEP);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_WIN_INIT_SEP);
+
+    finfo = fi_dupinfo(MPIDI_Global.prov_use);
+    MPIR_Assert(finfo);
+
+    /* Initialize scalable EP when first window is created. */
+    if (MPIDI_Global.rma_sep == NULL) {
+        /* Specify number of transmit context according user input and provider limit. */
+        MPIDI_Global.max_rma_tx_cnt = MPL_MIN(MPIDI_Global.prov_use->domain_attr->max_ep_tx_ctx,
+                                              MPIR_CVAR_CH4_OFI_MAX_RMA_SEP_CTX);
+        finfo->ep_attr->tx_ctx_cnt = MPIDI_Global.max_rma_tx_cnt;
+
+        MPIDI_OFI_CALL(fi_scalable_ep(MPIDI_Global.domain, finfo, &MPIDI_Global.rma_sep, NULL), ep);
+        MPIDI_OFI_CALL(fi_scalable_ep_bind(MPIDI_Global.rma_sep, &(MPIDI_Global.av->fid), 0), bind);
+        /* Allocate and initilize tx index array. */
+        utarray_new(MPIDI_Global.rma_sep_idx_array, &ut_int_icd, MPL_MEM_RMA);
+        for (i = 0; i < MPIDI_Global.max_rma_tx_cnt; i++) {
+            utarray_push_back(MPIDI_Global.rma_sep_idx_array, &i, MPL_MEM_RMA);
+        }
+    }
+    /* Set per window transmit attributes. */
+    MPIDI_OFI_set_rma_fi_info(win, finfo);
+    /* Get available transmit context index. */
+    int *index = (int *) utarray_back(MPIDI_Global.rma_sep_idx_array);
+    if (index == NULL) {
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+    /* Retrieve transmit context on scalable EP. */
+    MPIDI_OFI_CALL(fi_tx_context
+                   (MPIDI_Global.rma_sep, *index, finfo->tx_attr, &(MPIDI_OFI_WIN(win).ep), NULL),
+                   ep);
+    MPIDI_OFI_WIN(win).sep_tx_idx = *index;
+    /* Pop this index out of reserving array. */
+    utarray_pop_back(MPIDI_Global.rma_sep_idx_array);
+
+    MPIDI_OFI_CALL(fi_ep_bind(MPIDI_OFI_WIN(win).ep,
+                              &MPIDI_Global.ctx[0].cq->fid, FI_TRANSMIT | FI_SELECTIVE_COMPLETION),
+                   bind);
+    /* Check whether per-window counter is used. */
+    if (MPIDI_OFI_win_set_per_win_sync(win) == MPI_SUCCESS) {
+        MPIDI_OFI_CALL(fi_enable(MPIDI_OFI_WIN(win).ep), ep_enable);
+        goto fn_exit;
+    }
+    /* Fail in per-window sync, close opened tx. */
+    MPIDI_OFI_CALL(fi_close(&MPIDI_OFI_WIN(win).ep->fid), epclose);
+    utarray_push_back(MPIDI_Global.rma_sep_idx_array, &(MPIDI_OFI_WIN(win).sep_tx_idx),
+                          MPL_MEM_RMA);
+    mpi_errno = MPI_ERR_OTHER;
+    goto fn_fail;
+
+  fn_exit:
+    fi_freeinfo(finfo);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_WIN_INIT_SEP);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
```

Comment:
The failure handling in `MPIDI_OFI_win_init_sep` is unclear to me. In current code:
1. If `fi_scalable_ep()` fails, `MPIDI_Global.rma_sep` will be NULL and `rma_sep_idx_array` will be NULL.
2. If `fi_scalable_ep_bind()` fails, `rma_sep` will not be NULL but `rma_sep_idx_array` will be NULL. However, `rma_sep_idx_array` will be illegally accessed in a later window creation call.
3. If `utarray_new()` fails (unlikely), no failure handling at all.
4. If `index == NULL`, which means no available `tx_ctx`, it returns `MPI_ERR_OTHER`.
5. If `fi_tx_context()` fails, just return and no `index` is popped.
6. If `fi_ep_bind()` fails, just return but `index` is popped.
7. If `MPIDI_OFI_win_set_per_win_sync` or `fi_enable` fails, the opened tx will be closed and `index` will be pushed back.

The concerns are:
- In 1, the SEP creation will be re-tried in every later window. This might be OK since the SEP resource might become available at a later time ?
- 2 and 3 need to be fixed
- In 4, "no available tx" is a valid case, we should not set an MPI error code here (it is not handled by caller routine though). We might want to use a flag for a "valid failure" ?
- 6 needs to be fixed: we should push `index` back.

It will be better if the code can clearly define the "valid failure" (e.g., no available resource) and "system failure" cases. The **MPI_ERR** code should be used only at "system failure", which might cause undefined behavior at continued execution; all "valid failure" cases should be handled by runtime properly.

Reply:
if `utarray_new()` fails, it just exists with `-1`. no failure handling can be added here.